stdlib/asm/crypto/hashes/blake3.masm

```diff
@@ -1,579 +1,435 @@
     u32xor
     movdn.3
 
-    dup.12
-    movup.5
+    movup.4
+    movup.8
     u32xor
     movdn.4
 
-    dup.13
-    movup.6
+    movup.5
+    movup.8
     u32xor
     movdn.5
 
-    dup.14
-    movup.7
+    movup.6
+    movup.8
     u32xor
     movdn.6
 
-    dup.15
+    movup.7
     movup.8
     u32xor
     movdn.7
 end
 
-# column-wise mixing
-# see https://github.com/BLAKE3-team/BLAKE3/blob/da4c792d8094f35c05c41c9aeb5dfe4aa67ca1ac/reference_impl/reference_impl.rs#L55-L59
+# Given blake3 state matrix ( total 16 elements, each of 32 -bit ) and 
+# 8 message words ( each of 32 -bit ), this routine performs column-wise mixing
+# of message words into blake3 hash state.
+#
+# Functionality wise this routine is equivalent to https://github.com/BLAKE3-team/BLAKE3/blob/da4c792/reference_impl/reference_impl.rs#L55-L59
+#
+# Expected stack state:
+#
+# [state0_3_addr, state4_7_addr, state8_11_addr, state12_15_addr, m0, m1, m2, m3, m4, m5, m6, m7]
+#
+# Note, state_`i`_`j`_addr -> absolute address of {state[i], state[i+1], state[i+2], state[i+3]} in memory | j = i+3
+#
+# Meaning four consecutive blake3 state words can be read from memory easily.
+#
+# Final stack state:
+#
+# [state0, state1, state2, state3, state4, state5, state6, state7, state8, state9, state10, state11, state12, state13, state14, state15]
+#
+# i.e. whole blake3 state is placed on stack ( in order ).
 proc.columnar_mixing.1
-    pushw.mem
-    popw.local.0
+    swapw.2
+    swapw
 
-    pushw.local.0
-    drop
-    repeat.2
-        swap
-        drop
-    end
-    pushw.mem           # = b
+    movup.7
+    movup.6
+    movup.5
+    movup.4
 
-    pushw.local.0
-    repeat.3
-        swap
-        drop
-    end
-    pushw.mem           # = a
+    storew.local.0
 
-    dup.4
     movup.9
-    u32overflowing_add
-    drop
-    u32overflowing_add
-    drop
+    loadw.mem
+    movup.8
+    pushw.mem
 
-    dup.1
-    dup.6
-    movup.10
-    u32overflowing_add
-    drop
-    u32overflowing_add
-    drop
-    swap.2
+    movup.8
+    dup.5
+    u32unchecked_add3
     drop
 
-    dup.2
-    dup.7
-    movup.10
-    u32overflowing_add
-    drop
-    u32overflowing_add
-    drop
-    swap.3
+    swap
+    movup.8
+    dup.6
+    u32unchecked_add3
     drop
+    swap
 
-    dup.3
-    dup.8
-    movup.10
-    u32overflowing_add
-    drop
-    u32overflowing_add
-    drop
-    swap.4
+    movup.2
+    dup.6
+    movup.9
+    u32unchecked_add3
     drop
+    movdn.2
 
-    # -----
-
-    pushw.local.0
-    drop
-    drop
+    movup.3
+    dup.7
+    movup.9
+    u32unchecked_add3
     drop
-    pushw.mem           # = d
+    movdn.3
 
-    dupw.1              # copy a
+    movup.9
+    pushw.mem
 
-    movup.4
+    dup.4
     u32xor
     u32rotr.16
-    
+
     swap
-    movup.4
+    dup.5
     u32xor
     u32rotr.16
     swap
-    
+
     movup.2
-    movup.4
+    dup.6
     u32xor
     u32rotr.16
     movdn.2
-    
+
     movup.3
-    movup.4
+    dup.7
     u32xor
     u32rotr.16
```

Comment:
There is a refactoring in progress which will introduce `u32unchecked_rotr`. This will reduce the number of cycles required for this operation from 6 to 3. So, we should get some extra performance gains in the future.

Reply:
That's good. We use bit rotation quite a lot in cryptographic routines, so I see, many routines will benefit from that.